File: backend/src/security/audit.ts
import { Database } from "../infra/database.js";
import { redactJson } from "./redaction.js";

/** 将运行和安全事件脱敏后写入持久化运行事件表。 */
export class AuditWriter {
  /** 绑定唯一持久化事件入口，拒绝绕过脱敏直接写安全事件。 */
  constructor(private readonly database: Database) {}
  /** 序列化并脱敏事件元数据，再追加到数据库。 */
  async write(input: {
    traceId: string;
    eventType: string;
    result: string;
    metadata?: Record<string, unknown>;
  }): Promise<void> {
    const payload = redactJson({
      result: input.result,
      metadata: input.metadata ?? {},
    });
    this.database.appendEvent(input.eventType, input.traceId, payload);
  }
}

File: backend/app/security/audit.py
from __future__ import annotations

import json
from collections.abc import Mapping
from typing import Any

from app.infra.database import Database
from app.security.redaction import redact


class AuditWriter:
    """将运行和安全事件脱敏后写入持久化事件表。"""

    def __init__(self, database: Database) -> None:
        """绑定用于保存脱敏审计事件的数据库。"""
        self.database = database

    async def write(
        self,
        *,
        trace_id: str,
        event_type: str,
        result: str,
        metadata: Mapping[str, Any] | None = None,
    ) -> None:
        """序列化并脱敏事件元数据，再追加到数据库。"""
        payload = {
            "result": result,
            "metadata": metadata or {},
        }
        serialized = redact(json.dumps(payload, ensure_ascii=False, sort_keys=True, default=str))
        self.database.append_event(event_type, trace_id, serialized)
